Copy semantics/Trait for a_range::From

`::From` should have Copy semantics or Traits available (either derived or implemented by us). 

Since multiple ranges can have a single starting point, it makes sense for a user to be able to reuse a `::From` object.

An example:

```rust
let start = a_range::from(10);

let up_range = start.up_to(13); // [10, 11, 12, 13]

let down_range = start.down_to(8); // No copy semantics, so this fails to compile.
```

Formally, this is the kind of compile error I get right now:

```
error[E0382]: use of moved value: `x`
 --> src/lib.rs:27:18
  |
6 | let up_range = x.up_to(10);
  |                - value moved here
7 | let down_range = x.down_to(10);
  |                  ^ value used here after move
  |
  = note: move occurs because `x` has type `a_range::From<i32>`, which does not implement the `Copy` trait
```

Since `Idx` requires `Clone`, `From` can also derive `Clone`. I don't want `Idx` to only allow `Copy` types, though.

Copy is also kind of special IIRC. Maybe we can `impl<Idx> Copy for From when Idx: Copy`? Otherwise, this might be difficult.

> Am 19.10.2018 um 21:49 schrieb Abhishek Bhattacharya <notifications@github.com>:
> 
> ::From should have Copy semantics or Traits available (either derived or implemented by us).
> 
> Since multiple ranges can have a single starting point, it makes sense for a user to be able to reuse a ::From object.
> 
> An example:
> 
> let start = a_range::from(10
> );
> 
> 
> let up_range = start.up_to(13); // [10, 11, 12, 13]
> 
> 
> 
> let down_range = start.down_to(8); // No copy semantics, so this fails to compile.
> Formally, this is the kind of compile error I get right now:
> 
> error[E0382]: use of moved value: `x`
>  --> src/lib.rs:27:18
>   |
> 6 | let up_range = x.up_to(10);
>   |                - value moved here
> 7 | let down_range = x.down_to(10);
>   |                  ^ value used here after move
>   |
>   = note: move occurs because `x` has type `a_range::From<i32>`, which does not implement the `Copy` trait
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -65,6 +65,17 @@ pub struct From<Idx> {
     from: Idx,
 }
 
+/// Implement copy for From instances with copyable indices
+impl<T: Copy> Copy for From<T> {}
+
+#[test]
+fn copyable_from_is_copy() {
+    fn is_copy<T: Copy>(_x: T) {}
+
+    let range = From { from: 0 };
+    is_copy(range);
+}
+
 impl<Idx> From<Idx>
 where
     Idx: PartialOrd,
@@ -146,12 +157,27 @@ pub struct Range<Idx, Direction> {
     to: Idx,
 }
 
+/// Implement copy for Ranges with copyable indices
+impl<T: Copy, Direction: Copy> Copy for Range<T, Direction> {}
+
+#[test]
+fn copyable_range_is_copy() {
+    fn is_copy<T: Copy>(_x: T) {}
+
+    let range = Range {
+        direction: Upwards,
+        from: 0,
+        to: 100,
+    };
+    is_copy(range);
+}
+
 #[doc(hidden)]
-#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd)]
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd)]
 pub struct Upwards;
 
 #[doc(hidden)]
-#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd)]
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd)]
 pub struct Downwards;
 
 /// Range counting up

